### PR TITLE
Add full scoping support for Library

### PR DESCRIPTION
**Pull request by matanlurey**
Also ran dartfmt and organized members by name in classes so the diff is much larger than it actually is (sorry about that). Added more tests for when a `Scope` is used.

### CODE REVIEW

File: lib/code_builder.dart [hunk 1]
```diff
@@ -44,60 +44,47 @@ part 'src/builders/method_builder.dart';
...
 part 'src/scope.dart';
 
+final DartFormatter _dartfmt = new DartFormatter();
+
+// Simplifies some of the builders by having a mutable node we clone from.
+/// Returns [source] formatted by `dartfmt`.
+@visibleForTesting
+String dartfmt(String source) => _dartfmt.format(source);
+
+// Creates a defensive copy of an AST node.
```
Comment (yjbanov) [inline lib/code_builder.dart]: Is cloning only useful for defensive purposes? If not, I'd drop "defensive". Also, if the cloning is deep, I'd note that.

File: lib/code_builder.dart [hunk 2]
```diff
@@ -44,60 +44,47 @@ part 'src/builders/method_builder.dart';
...
+AstNode/*=E*/ _cloneAst/*<E extends AstNode>*/(AstNode/*=E*/ astNode) {
+  return new AstCloner().cloneNode/*<E>*/(astNode);
+}
+
+Identifier _stringId(String s) {
+  return new SimpleIdentifier(new StringToken(TokenType.STRING, s, 0));
+}
+
+Literal _stringLit(String s) {
```
Comment (yjbanov) [inline lib/code_builder.dart]: dart style police calls: avoid abbreviations

File: lib/code_builder.dart [hunk 3]
```diff
@@ -44,60 +44,47 @@ part 'src/builders/method_builder.dart';
...
 abstract class CodeBuilder<A extends AstNode> {
   /// Returns a copy-safe [AstNode] representing the current builder state.
-  A toAst();
+  ///
+  /// Uses [scope] to output an AST re-written to use appropriate prefixes.
+  A toAst([Scope scope = const Scope.identity()]);
 }
 
-// Simplifies some of the builders by having a mutable node we clone from.
+@Deprecated('Builders are all becoming lazy')
```
Comment (yjbanov) [inline lib/code_builder.dart]: Is it worth deprecating an internal class?

File: lib/src/builders/class_builder.dart [hunk 4]
```diff
@@ -5,79 +5,96 @@
...
-    }
-    return new ClassBuilder._(astNode);
-  }
+    TypeBuilder extend,
+    Iterable<TypeBuilder> implement: const [],
+    Iterable<TypeBuilder> mixin: const [],
+  }) =>
+      new ClassBuilder._(
+        name,
+        abstract,
```
Comment (yjbanov) [inline lib/src/builders/class_builder.dart]: While legal Dart code I'm not a fan of Dart keywords used as identifiers, especially in the public API (this named parameter is visible to user code)

File: lib/src/builders/expression_builder.dart [hunk 5]
```diff
@@ -93,10 +109,10 @@ abstract class ExpressionBuilder implements CodeBuilder<Expression> {
   }); // TODO(matanl): Rename to invoke when factory is removed.
 
   /// Returns wrapped as a [ExpressionFunctionBody] AST.
-  ExpressionFunctionBody toFunctionBody() => _asFunctionBody(this);
+  ExpressionFunctionBody toFunctionBody([Scope scope = const Scope.identity()]) => _asFunctionBody(this, scope);
 
   /// Returns wrapped as a [FunctionExpression] AST.
-  FunctionExpression toFunctionExpression() => _asFunctionExpression(this);
+  FunctionExpression toFunctionExpression([Scope scope = const Scope.identity()]) => _asFunctionExpression(this, scope);
```
Comment (yjbanov) [inline lib/src/builders/expression_builder.dart]: This is a looooong line. Did you dartfmt this file?

File: lib/src/builders/field_builder.dart [hunk 6]
```diff
@@ -30,11 +31,13 @@ class FieldBuilder implements CodeBuilder<Declaration> {
     this._name, {
     TypeBuilder type,
     ExpressionBuilder initialize,
+    bool static: false,
```
Comment (yjbanov) [inline lib/src/builders/field_builder.dart]: Another Dart keyword

File: lib/code_builder.dart (hunk 1 above)
Comment (matanlurey) [inline lib/code_builder.dart]: defensive --> deep.

Done.

File: lib/code_builder.dart (hunk 2 above)
Comment (matanlurey) [inline lib/code_builder.dart]: Done.

File: lib/code_builder.dart (hunk 3 above)
Comment (matanlurey) [inline lib/code_builder.dart]: It's helping me (in the analyzer) show what work I have left to do

File: lib/src/builders/class_builder.dart (hunk 4 above)
Comment (matanlurey) [inline lib/src/builders/class_builder.dart]: Done.

Split off to ClassBuilder.asAbstract(...)

File: lib/src/builders/expression_builder.dart (hunk 5 above)
Comment (matanlurey) [inline lib/src/builders/expression_builder.dart]: Done.

File: lib/src/builders/field_builder.dart (hunk 6 above)
Comment (matanlurey) [inline lib/src/builders/field_builder.dart]: --> asStatic. Thanks!

**Pull request merged**